Adds MYHPOM backed IRODs implementation (MH-76)

dsummersl opened the pull request:
This PR extends hydroshare's django storage engine implementation to use IRODS as the storage mechanism. Rather than storing files locally, they should be uploaded to a configured IRODS server.

To verify that a file has been uploaded you can use the IRODS i* commands (much like a CLI FTP client) to see/obtain files in its system.

Notes:
 * To minimize the effort for this feature I built on top of the django_irods application's implementation:
    * MyhpomStorage provides slight modifications to how the default worked.
    * The download view is based upon the django_irods download view, and customized for our use cases.
 * This PR is intended to work by default without irods configured - you should see the same behavior in the UI for downloading/uploading AD templates and ADs, but different behavior on the filesystem:
    * FileSystemStorage: files are uploaded to `hydroshare/public/media`, and download links will be provided by the default storage engine (`/media/`).
    * MyhpomStorage: files are not present anywhere on the local filesystem, are only visible via the irods, and download links will be that provided in the myhpom app (`/download/`).
 * When the irods servers are set up, all media will be stored/fetched from the irods servers. When you switch your local dev environment between storage methods you will need to re-upload AD templates, and ADs in order for them to be found.
 * When this PR is deployed we will need to re-upload the AD templates, and delete any test ADs we've been using.

To test this branch you need to test in two modes: without irods (what you've been doing in local dev mode), and with irods. To set up your local dev environment to use irods (see the `irods/README` for more details):

```bash
docker-compose stop
cd irods
docker login -u USERNAME -p PASSWORD
./use-local-irods.sh
cd ..
./hsctl rebuild
```

The `use-local-irods.sh` script will start two additional docker containers (IRODs servers), and configure your local_settings.py file and hsctl file to use them. They are not a part of your `docker-compose` file, however, so use `docker ps` to see them.

Once your django services are running, upload an AdvanceDirective. You can then use the hydroshare images to issue IRODS commands and see the files that have been uploaded:

```bash
# Create a new hydroshare container (which has all the i* commands):
docker-compose run --rm hydroshare bash

# Setup the local i* commands with credentials to access the irods servers:
iinit
# You will be prompted for several settings: host/username/password/etc.
# Consult your locally modified local_settings.py file for the corresponding IRODS_ commands

# View all directories in irods:
ils
# example output:
/hydroshareZone/home/wwwHydroProxy:
  C- /hydroshareZone/home/wwwHydroProxy/myhpom

ils /hydroshareZone/home/wwwHydroProxy/myhpom
/hydroshareZone/home/wwwHydroProxy/myhpom:
  ad.pdf
  C- /hydroshareZone/home/wwwHydroProxy/myhpom/advance_directives

# try to fetch some data:
iget -r /hydroshareZone/home/wwwHydroProxy/myhpom/advance_directives

# and there it is:
ls advance_directives
newname_5j2KAU2.pdf
```

---

 * [x] Write test cases for MyhpomStorage. This might be nothing more than verifying certain sequences of irods commands are passed to the Session.
 * [x] Write test cases for irods_download view.
 * [x] Verify that files are only created in irods, and not in the media bucket (or elsewhere).
 * [x] Manually test deleting a record.
 * [x] Manually test removing a record - is the old record removed in irods?

dsummersl committed:
Adds MYHPOM backed IRODs implementation (MH-76)

dsummersl committed:
Merge branch 'myhpom-develop' into mh-76-add-irods

dsummersl committed:
Don't use MyhpomStorage when running tests (MH-76)

dsummersl committed:
Implements irods storage and adds test cases (MH-76)

dsummersl committed:
Fix non-irods tests & configure deployment environments (MH-76)

dsummersl committed:
Setup local irods to use MyhpomStorage (MH-76)

dsummersl committed:
Have MyhpomStorage open files in the browser (MH-76)

dsummersl committed:
Clean up comments (MH-76)

dsummersl committed:
Removes unnecessary test setting (MH-76)

dsummersl committed:
Merge branch 'myhpom-develop' into mh-76-add-irods

sharrisoncaktus committed:
[mh-76] edited the irods/README.md to reflect that docker login needs to have succeeded before use-local-irods.sh is run. Also spellign

sharrisoncaktus commented on myhpom/storage.py:
```diff
@@ -0,0 +1,22 @@
+        # Use the default method of creating the name used to save 'name' as.
+        # The IrodsStorage implementation will prevent files with the same
+        # 'name' from being stored:
+        return Storage.get_available_name(self, name)
```
My concern here is that, if a user happens to be familiar with Django, and their AD is named MYAD_74Y5tX9.pdf, then they are going to know that there is **probably** a file named MYAD.pdf, which belongs to the first person who uploaded a file named MYAD.pdf. So it would be better for the get_available_name() method to **always** use something random that can be guaranteed unique. For example, using md5 to append a 16-byte hash of the filename + username + timestamp, or of a uuid.uuid4().

I realize this also should include testing the irods filesystem for the existence of that generated filename, which is a pain.

dsummersl replied:
That is the default behavior of django storage classes...I guess because most projects don't keep a unique filename around for their FileField and django decided to do the minimum necessary to prevent file clashes without annoying developers by always giving them funny looking filenames? 

I see a couple options (and I'm not sure what makes the most sense):
 * add additional filename obfuscation here...it means we need to also add an `original_filename` type field for `State.advance_directive_template` since the name that users download will now also be 'funny looking'. It also means that our local dev environments won't act quite the same as the production environments (b/c this MyhpomStorage will behave subtly differently than FileSystemStorage).
 * keep this drive as it is, so that it acts the same as other storage drivers -- but for files that matter (AdvanceDirective) rename the uploaded file in an obfuscated way.
 * don't do anything :/

I think I'm partial to...the second option?

sharrisoncaktus replied:
On one hand, if we do the fix below and there are never files left lying around, then the security hole is never hit. 

But I think I would be most comfortable with the second option — always append a random hash or at least short string to the basename ... at least for users' advance directives?

dsummersl replied:
Okay, I'll do that - I realize part of #1 will have to happen anyway...because eventually Gerald *will* notice there are filename clashes for state advance directives. I guess I'll make a task in the backlog for that.

dsummersl replied:
Created MH-177 for ^

dsummersl committed:
Upload AD templates into their own directory (MH-76)

Now that there is a separate directory for ADs it makes sense to keep
the templates in their own directory too.

sharrisoncaktus commented on myhpom/tests/test_irods_view.py:
```diff
@@ -0,0 +1,84 @@
+        advancedirective.user = another_user
+        advancedirective.save()
+        response = self.client.get(self.url)
+        self.assertEqual(404, response.status_code)
```
This test passes on my system, BUT: If the user deletes their advance directive, it leaves it on the filesystem. So the above comment is still valid: 

1. User 1 uploads MYAD.pdf, which is saved with that filename. 
2. User 2 uploads MYAD.pdf, which is saved as MYAD_RANDOM.pdf.
3. User 1 "deletes" their AD, but MYAD.pdf remains in the irods filesystem. 
4. Now User 2, who sees MYAD_RANDOM.pdf, can download MYAD.pdf.

dsummersl replied:
Good catch!

From what I can tell its not the responsibility of the Storage drivers to go back and delete a file when a model is deleted (https://docs.djangoproject.com/en/1.11/ref/models/fields/#django.db.models.fields.files.FieldFile.delete -- which makes some sense...). I've heard of two ways around this in django:

 * make a cron job that, for each model with a FileField you care about, computes the set of files that should exist...and then delete any that aren't in that set. Slow, bet reliable.
 * use pre_save/post_delete hooks on the models to delete the file. Fast, but could fail for bulk operations (if we used them).

I suggest...I'm not sure? The second option (add pre_save/post_delete hooks to AdvanceDirective)?

sharrisoncaktus replied:
I think I would do option 2 and then, if we ever implement bulk ops, swap it to the more complicated option 1.

dsummersl replied:
Okay, I'll start with that then.

sharrisoncaktus reviewed (changes requested):
It seems we have a security hole – see comments.

sharrisoncaktus committed:
Merge branch 'mh-76-add-irods' of https://github.com/myhpom/MyHPOM into mh-76-add-irods

dsummersl committed:
Ensure AD filenames aren't guessable (MH-76)

dsummersl committed:
Deletes underlying documents when the AD is deleted (MH-76)

dsummersl committed:
Increases number of random digits on a file (MH-76)

dsummersl committed:
Merge branch 'myhpom-develop' into mh-76-add-irods

dsummersl committed:
Merge branch 'mh-76-add-irods' of github.com:myhpom/MyHPOM into mh-76-add-irods

sharrisoncaktus reviewed (approved):
Code changes look good and app works great. J'approuve!
